### PR TITLE
Switchboard Jobs Part 2

### PR DESCRIPTION
**Pull request by JackTemaki**
As continuation of #109 now the following additions:

 - A job to convert dual channel sphere files to single channel wav
 - 3 Jobs to prepare the 3 different datasets, as they are all slightly different (but share code)
 - A job to prepare the "BPE" style corpus for e.g. setups of @albertz  and @mmz33, which uses extra mapping files that were somehow created by Kazuki and are non reproducable for me, thus I wanted to add them directly.
 
 PR is a draft as not everything is fully tested yet.
 
 As I am probably the least familiar person with Switchboard, I really encourage everyone to have a look.
 
 You can find the produced eval set files for review here:
  - Hub5e00: `/u/rossenbach/experiments/tts_asr_2021/output/swbd_hub5e_00_testing`
  - Hub5e01: `/u/rossenbach/experiments/tts_asr_2021/output/swbd_hub5e_01_testing`
  - RT03s: `/u/rossenbach/experiments/tts_asr_2021/output/swbd_rt03s_testing`
  
  - New training corpus with new audio: `/u/rossenbach/experiments/tts_asr_2021/output/Switchboard-LDC/train.corpus.gz`
  - New training corpus with legacy audio: `/u/rossenbach/experiments/tts_asr_2021/output/Switchboard-i6-legacy/train.corpus.gz` 
  
  -  For the BPE cleaning compare in `/u/rossenbach/experiments/tts_asr_2021/output/swbd_bpe_test` the two files `corpus.txt` and `corpus_cleaned.txt`

### CODE REVIEW

**Comment by albertz** [comment]
Regarding the number normalization, we have `normalize_numbers` in RETURNN, or `english_cleaners`, which does sth very similar as what you have in your mapping. It uses the `inflect` library: https://github.com/jaraco/inflect

**Review by michelwi — COMMENTED** [review]
I like your very well documented and clear Jobs.

File: datasets/switchboard.py [hunk 1]
```diff
@@ -336,3 +339,337 @@ def run(self):
...
+            )
+
+
+#### Evaluation Corpus Helper ####
+
+
+def _process_and_write_stm(stm_in_files: List[str], stm_out_file):
+    """
+    Kaldi-preprocessing (remove double brackets, remove <B_ASIDE> and <E_ASIDE>
+    Change naming pattern to Zoltan style
```
[the remark this answers is not in the document]
Comment (JackTemaki) [reply datasets/switchboard.py]: Maybe Zoltan style is the wrong word, but we need to alter the names a little bit to fit with the bliss. I really do not want to touch that. I updated the documentation accordingly

File: datasets/switchboard.py [hunk 2]
```diff
@@ -336,3 +339,325 @@ def run(self):
...
+        yield Task("run", rqmt=self.rqmt)
+
+    def run(self):
+        for sph_file in glob.glob(
+            os.path.join(self.sph_audio_folder.get_path(), "**/*.sph"), recursive=True
+        ):
+            sph_name, ext = os.path.splitext(os.path.basename(sph_file))
+            subprocess.call(
+                [
+                    "ffmpeg",
```
Comment (Marvin84) [inline datasets/switchboard.py:370]: why not `sox`? Do we know what differs here?
I think Zoltan's original script was using `sox` and also AppTek for `raw->wav` generally uses it.

Comment (JackTemaki) [reply datasets/switchboard.py:370]: I do not want to add an additional dependency to the pipeline, everything else works with FFmpeg.
I am more familiar with FFmpeg, and the results do not (or rather, can not) differ (Peter and I checked), as this is numerically stable conversion from mulaw-pcm to linear-pcm.

**Review by michelwi — APPROVED** [review]

**Review by Marvin84 — APPROVED** [review]

File: datasets/switchboard.py [hunk 3]
```diff
@@ -336,3 +339,348 @@ def run(self):
...
+                ]
+            )
+
+
+#### Evaluation Corpus Helper ####
+
+
+def _process_and_write_stm(stm_in_files: List[str], stm_out_file: str):
+    """
+    Kaldi-preprocessing (remove double brackets, remove <B_ASIDE> and <E_ASIDE>
```
Comment (Atticus1806) [inline datasets/switchboard.py:396]:
```suggestion
    Kaldi-preprocessing (remove double brackets, remove <B_ASIDE> and <E_ASIDE>)
```

File: datasets/switchboard.py [hunk 4]
```diff
@@ -336,3 +339,348 @@ def run(self):
...
+
+class SwitchboardSphereToWaveJob(Job):
+    """
+    Takes an audio folder from one of the switchboard LDC folders and converts dual channel .sph files
+    with mulaw encoding to single channel .wav files with s16le encoding
+    """
+
+    def __init__(self, sph_audio_folder: tk.Path):
+        """
+        :param sph_audio_folder:
```
Comment (Atticus1806) [inline datasets/switchboard.py:352]: Just for completeness a small doc pls. :)

**Comment by JackTemaki** [comment]
I will merge this now, and then add another PR which includes the Fisher transcriptions and some refactoring.

**Pull request merged**